diff --git a/services/vision_msgs/msg/cop_answer.msg b/services/vision_msgs/msg/cop_answer.msg
index d58a0e1..d753e9f 100644
--- a/services/vision_msgs/msg/cop_answer.msg
+++ b/services/vision_msgs/msg/cop_answer.msg
@@ -1,3 +1,4 @@
 #Message that cop uses to  answer on specified topics
 string error                        # eventuelly there are errors like no object found
 aposteriori_position[] found_poses  # list od found pose candidate for a certain query
+uint64 callId
\ No newline at end of file
